### DISCUSSION TITLE
Documentation for in-fight AI system

### DISCUSSION THREAD

**Discussion by ghost** [ihhub/fheroes2#5717] [General]
### Your question

Hi. I'm interested at improving the in-fights AI and would like to have some guidance at where to look in codebase. Ideally, and IRC channel or something I can reach with a bridge, and some documentation, even WiP.

Combat AI still have the old HoMM2 problems imo, at it still does basic tactical mistakes the child I was could consider stupid back in time.
I have not played this since decades, except for the few games I had those last weeks, so I'm likely missing tactics, but it's still too easy, so I think I should be able to fix various issues. This particular side of AI would merit it's own tag, it would help if I really go in hacking this.

### Additional info

When looking at open issues, I see those are related to this problem:

* #2729 
* #2750
* #2751
* #2954
* #3380 
* #3393 
* #3416
* #3497 
* #3508 
* #3905
* #4124 
* #4129 
* #4130 
* #4148 
* #4254 
* #4382 
* #4585 
* #5144 
* #5147 
* #5325 
* #5515 
* #5523 
* #5515 
* #5523 
* #5648

**Comment by ihhub** [comment]
Hi @bmorel , you can look at **ai_normal_battle.cpp** file for AI action during battle. You can also join our Discord server for discussions about the development.

**Comment by ihhub** [comment]
I am moving this issue into discussion section.

**Comment by ghost** [comment]
Just placing a message here to be noticed if more events happen on this. Looking at source code is not really a good way to learn about architecture and concepts.
Myself, I'm starting to be pretty familiar with FPS bots AI mechanisms (expert systems),  but not about the mechanisms that need to be used to drive other entities. On that point, an interesting document that *might* apply partially to HoMM2 is https://www.researchgate.net/publication/240430519_The_Quake_III_Arena_Bot, since it seems they did implemented some kind of "leader" AI, but it was still giving instructions to autonomous AIs.

When I say I'm starting to be familiar, it's just because I've been hacking existing bots for a FOSS FPS game for around 2 years now, with pretty decent results, not that I have any academic knowledge on the matter.

**Reply by ihhub** [reply]
Hi @bmorel ,

writing documentation is out of the scope of this project so the core team members will not do this as they have much higher priority tasks. AI in Heroes 2 is way more complex than in FPS games due to the scale of logic in the game itself. The best way to understand how it works is actually to read the source code.
